### ISSUE TITLE
Borrow checker rule violation in VisitMut trait

### ISSUE THREAD

**Issue by rzvxa** [oxc-project/oxc#2745]
Hello guys,

I was going through the source of the project when I found this that may cause problems in the future if used with wrong assumptions.
My understanding is that there are 2 types of visitors in this project, `Visit` and `VisitMut`, These would take references to AST items and transmute them to the desired lifetime using the `self.alloc` method on both of these traits.
This part is scary but could be fine, The issue comes down to how it behaves in the `VisitMut` trait, It not only transmutes the lifetime but because of implicit cast happening in the reinterpretation, it will create an immutable reference without letting the borrow checker know about it which is illegal since we are also borrowing the items as mutable for traversing the tree.

Here's a usage example from the code.
```rust
    fn visit_block_statement(&mut self, stmt: &mut BlockStatement<'a>) {
        let kind = AstKind::BlockStatement(self.alloc(stmt)); // immutable borrow happens here
        self.enter_scope(ScopeFlags::empty());
        self.enter_node(kind); // passing the immutable borrow to an external struct implementation which would store it
        self.visit_statements(&mut stmt.body); // mutable borrow which should be illegal since both internal oxc crates in future and third-party tools may rely on it not changing.
        self.leave_node(kind);
        self.leave_scope();
    }
```

In case I'm mistaken about this issue please let me know but if my understanding is correct this could cause undefined behaviors further down the road.

I think we shouldn't need to pass the item with the AstKind enum, we can split it into an enum storing only the type and a union for passing the item, With this approach we should be able to pass the union when entering a node and only pass the kind when leaving it which as it is right now would serve the same purpose since we always ignore the value in AstKind when leaving the node(we are just acting based on the type).

I would love to work on this issue if you guys find it relative and the suggested implementation appropriate, However, I'm open to doing something entirely different or letting the oxc team members take over the issue.

**Comment by Boshen** [comment]
Can you elaborate on "we can split it into an enum storing only the type and a union for passing the item", what does it look like?

The whole purpose of the `VisitMut` trait is to get the transformer working. Everything is a hack right now because I don't know what the correct pattern should be ... as long as the transformer works 😅 It'll probably break apart when more features are added ...

**Comment by rzvxa** [comment]
Hey, Thanks for the quick response,

> Can you elaborate on "we can split it into an enum storing only the type and a union for passing the item", what does it look like?
 
I was thinking that instead of having AstKind while visiting nodes, We could create an AstRef which would contain all the necessary information to use the said value. Then we can choose which information we are willing to share with other functions.
So instead of having this:

```rust
pub enum AstKind<'a> {
    Program(&'a Program<'a>),
    Directive(&'a Directive),
    Hashbang(&'a Hashbang),
    // ...
}
```

We could create a more verbose representation which should in theory perform the same since in both cases we only need a match to retrieve our reference.

```rust
pub union AstKind<'a> {
    program: &'a Program<'a>,
    directive: &'a Directive,
    hashbang: &'a Hashbang,
    // ...
}

pub enum AstType<'a> {
    Program,
    Directive,
    Hashbang,
    // ...
}

pub struct AstRef<'a> {
    kind: AstKind<'a>,
    r#type: AstType<'a>,
}
```

Now with this, we can rewrite our visitors like so:

```rust
fn visit_block_statement(&mut self, stmt: &'a mut BlockStatement<'a>) {
    let node = AstRef::new(AstKind { block_statement: &stmt }, AstType::BlockStatement);
    self.enter_scope(ScopeFlags::empty());
    self.enter_node(node); // passing immutable reference here.
    self.visit_statements(&mut stmt.body); // using mutable reference here.
    self.leave_node(node.r#type); // we no longer pass immutable reference after creating the mutable one, we just announce the leave event using the type.
    self.leave_scope();
}
```

#### *notice that I'm declaring the `stmt` parameter with the `a` lifetime but we still can use the `alloc` method for that reason if necessary.

Creating the `AstRef` can be done using the constructor, `From` trait, or even better a declarative macro that would map identifiers to both the `AstKind`, `AstType` and the necessary traits to initialize one.

I don't have much knowledge of different parts of the project so it could be wrong and unsuitable for the crates that I'm not familiar with.

**Comment by rzvxa** [comment]
I just wanted to add that if you want a minimal amount of refactoring it is also possible to achieve the same thing with something like this:
```rust
pub enum AstKind<'a> {
    Program(&'a Program<'a>),
    Directive(&'a Directive),
    Hashbang(&'a Hashbang),
    // ...
}

pub enum AstType<'a> {
    Program,
    Directive,
    Hashbang,
    // ...
}

impl AstKind {
    fn ast_type(&self) -> AstType {
        match self {
            Self::Program(_) => AstType::Program,
            Self::Directive(_) => AstType::Directive,
            Self::Hashbang(_) => AstType::Hashbang,
        }
    }
}
```

But I find the first approach more convenient since it doesn't rely on compiler optimization to get rid of the extra match, With the way mentioned above we have to do pattern matching twice once for getting the type and once for acting on the type.

```rust
let t = kind.ast_type(); // match #1
match t { // match #2
    // ...
}
```

**Comment by rzvxa** [comment]
Well, sadly this seems to not be a feasible approach.
https://github.com/oxc-project/oxc/blob/95ac2655047ab0692337cc1059b01f56b4ba76d8/crates/oxc_transformer/src/es2015/new_target.rs#L34C1-L38C6

The `new_target` transformer relies on the ast kind sent when leaving a node.
I think the cleanest way to do this is just getting rid of the `enter_node` and `leave_node` altogether. While it may be more cumbersome to do the same thing these methods do so by implementing every necessary method instead of a simple match, In the long run, it is much more predictable and even extendable at the cost of the nice abstraction we currently have around them.

[Similar to rust visitors](https://github.com/rust-lang/rust/blob/5608c7f9aaf380bd0c49bbcc350ecf8c5eb4b68c/compiler/rustc_ast/src/mut_visit.rs) we can also separate the visit methods and walk functions, This way even when we implement every method the visitor offers it still keeps the general abstraction via the walk functions.

**Comment by Boshen** [comment]
It's a dilemma. If we start changing things right now, I'm not sure if the transformer will still work as intended. So to be safe, I'll revisit this in probably another month (I'm currently busy doing some real-life things.)

Adding walk methods should be done at some point, but we should probably use a code generator to generate the whole file instead of updating it manually ... but I also want to manage the scope logic as well.

Rust is really getting in the way of doing this when it comes to mutating the ast :-(

**Comment by Boshen** [comment]
enter / leave node will be used to create a parent pointing tree, how are we going to achieve that without them?

> I was wondering how we test transformers. Is there a lack of tests for that crate? since I couldn't find tests in it and didn't see any relative tests on other crates.

The test suite is here: https://github.com/oxc-project/oxc/tree/main/tasks/transform_conformance

---

I think we should start from somewhere small, maybe override a few visitor functions in the transformer's `VisitMut` implementation and see what happens?

**Comment by rzvxa** [comment]
> enter / leave node will be used to create a parent pointing tree, how are we going to achieve that without them?
Instead of calling it from trait default implementation, we deal with it in each node visitor.

So my idea is that instead of having this in the semantic builder's Visit trait implementation:
```rust
    // Setup all the context for the binder.
    // The order is important here.
    fn enter_node(&mut self, kind: AstKind<'a>) {
        self.create_ast_node(kind);
        self.enter_kind(kind);
    }
```

We could have the same method in the SemanticBuilder's implementation itself`, And by having this we don't enforce it for every visitor, since it is fine to pass immutable references around in the case of SemanticBuilder, Where it is not ok we have to find a workaround but it would localize our issues to individual visitors(even if we do transmute to workaround it for now).

Despite it sounding a bit different, The actual change is going to be minimal, However, It would enforce the implementation of every method, which is where walk functions come into place.

For most visitors other than SemanticBuilder which is very dependent on its order of execution, we can just implement every visit method needed for our operations.

Take a simple visitor like `ASTPass` for example:

```rust
impl<'a> Visit<'a> for ASTPass {
    fn enter_node(&mut self, kind: AstKind<'a>) {
        match kind {
            AstKind::Function(_) => {
                self.number_of_functions += 1;
            }
            AstKind::Class(_) => {
                self.number_of_classes += 1;
            }
            _ => {}
        }
    }
}
```

We have to rewrite it like this to achieve the same effect:

```rust
impl<'a> Visit<'a> for ASTPass {
    fn visit_function(&mut self, func: &oxc_ast::ast::Function<'a>, flags: Option<oxc_syntax::scope::ScopeFlags>) {
        self.number_of_functions += 1;
        walk_function(self, func, flags)
    }

    fn visit_class(&mut self, class: &oxc_ast::ast::Class<'a>) {
        self.number_of_classes += 1;
        walk_class(self, class)
    }
}
```
